File: skills/agentforge/scripts/agentforge_api.py
#!/usr/bin/env python3
"""
AgentForge CLI - Command-line interface for AgentForge API

Usage:
  agentforge_api.py --method create --prompt "..." [options]
  agentforge_api.py --method list [--status STATUS]
  agentforge_api.py --method get --task-id ID
  agentforge_api.py --method runs --task-id ID
  agentforge_api.py --method output --task-id ID
  agentforge_api.py --method events --task-id ID [--limit N]
  agentforge_api.py --method cancel --task-id ID
  agentforge_api.py --method retry --task-id ID
  agentforge_api.py --method delete --task-id ID
  agentforge_api.py --method dag --dag-id ID
  agentforge_api.py --method health

Examples:
  agentforge_api.py --method create --prompt "Run tests" --title "Test run" --dir /path/to/project
  agentforge_api.py --method create --prompt "Backup" --schedule delayed --delay 300
  agentforge_api.py --method create --prompt "Deploy" --schedule scheduled_at --at "2026-02-15 14:30:00"
  agentforge_api.py --method create --prompt "Daily report" --schedule cron --cron "0 2 * * *"
  agentforge_api.py --method create --prompt "Step 2" --depends-on 10,11 --inject-result --dag-id my-pipeline
  agentforge_api.py --method list --status running
  agentforge_api.py --method get --task-id 5
  agentforge_api.py --method output --task-id 5
  agentforge_api.py --method retry --task-id 5
  agentforge_api.py --method delete --task-id 5
  agentforge_api.py --method dag --dag-id my-pipeline
"""

import argparse
import json
import sys
import os
import urllib.request
import urllib.error
import urllib.parse
from datetime import datetime


API_BASE = "http://127.0.0.1:9712/api"

SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))


def api_request(method, path, json_data=None, params=None):
    """Make HTTP request to AgentForge API using stdlib only."""
    url = f"{API_BASE}{path}"
    if params:
        url += "?" + urllib.parse.urlencode(params)

    body = None
    headers = {}
    if json_data is not None:
        body = json.dumps(json_data).encode("utf-8")
        headers["Content-Type"] = "application/json"

    req = urllib.request.Request(url, data=body, headers=headers, method=method)

    try:
        with urllib.request.urlopen(req) as resp:
            return json.loads(resp.read().decode("utf-8"))
    except urllib.error.URLError as e:
        if isinstance(e, urllib.error.HTTPError):
            try:
                err_body = json.loads(e.read().decode("utf-8"))
                detail = err_body.get("error", e.reason)
            except Exception:
                detail = e.reason
            print(f"API error (HTTP {e.code}): {detail}", file=sys.stderr)
        elif "Connection refused" in str(e.reason) or "No connection" in str(e.reason):
            print(
                "Connection failed: AgentForge is not running.\n"
                "Start it with: cd <project-root> && uv run taskboard.py",
                file=sys.stderr,
            )
        else:
            print(f"Network error: {e.reason}", file=sys.stderr)
        sys.exit(1)


STATUS_ICONS = {
    "pending": "🕐",
    "scheduled": "⏰",
    "running": "⏳",
    "completed": "✅",
    "failed": "❌",
    "cancelled": "🚫",
    "blocked": "⊘",
}


def main():
    parser = argparse.ArgumentParser(
        description="AgentForge CLI - Manage AI agent tasks",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog=__doc__,
    )

    parser.add_argument(
        "--method",
        required=True,
        choices=[
            "create", "list", "get", "runs", "output", "events",
            "cancel", "retry", "delete", "dag", "health",
        ],
        help="API method to call",
    )

    # Task creation parameters
    parser.add_argument("--prompt", help="Task prompt/instruction (for create)")
    parser.add_argument("--title", help="Task title (for create)")
    parser.add_argument("--dir", default=".", help="Working directory (for create, default: .)")
    parser.add_argument(
        "--schedule",
        choices=["immediate", "delayed", "scheduled_at", "cron"],
        default="immediate",
        help="Schedule type (for create)",
    )
    parser.add_argument("--cron", help="Cron expression (for create with schedule=cron)")
    parser.add_argument("--delay", type=int, help="Delay in seconds (for create with schedule=delayed)")
    parser.add_argument("--at", help="Specific datetime (for create with schedule=scheduled_at, format: 'YYYY-MM-DD HH:MM:SS')")
    parser.add_argument("--max-runs", type=int, help="Max runs for cron tasks (for create)")
    parser.add_argument("--tags", default="", help="Comma-separated tags (for create)")
    parser.add_argument("--image-paths", help="Comma-separated list of image file paths (for create)")
    # DAG parameters
    parser.add_argument("--depends-on", help="Comma-separated upstream task IDs (for create), e.g. '10,11'")
    parser.add_argument("--inject-result", action="store_true", help="Inject upstream results into prompt (for create with --depends-on)")
    parser.add_argument("--dag-id", help="DAG workflow group label (for create or dag method)")

    # Query parameters
    parser.add_argument("--task-id", type=int, help="Task ID (for get/runs/output/events/cancel/retry/delete)")
    parser.add_argument("--status", help="Filter by status (for list)")
    parser.add_argument("--limit", type=int, default=1000, help="Max events to return (for events)")
    parser.add_argument("--offset", type=int, default=0, help="Events to skip (for events)")

    # Output formatting
    parser.add_argument("--json", action="store_true", help="Output as JSON (for list/output)")

    args = parser.parse_args()

    # Route to appropriate handler
    if args.method == "create":
        if not args.prompt:
            parser.error("--prompt is required for create method")

        data = {
            "prompt": args.prompt,
            "title": args.title or args.prompt[:60],
            "working_dir": args.dir,
            "schedule_type": args.schedule,
            "tags": args.tags,
        }

        if args.cron:
            data["cron_expr"] = args.cron
        if args.delay:
            data["delay_seconds"] = args.delay
        if args.at:
            try:
                dt = datetime.strptime(args.at, "%Y-%m-%d %H:%M:%S")
                data["next_run_at"] = dt.isoformat()
            except ValueError:
                print("Error: --at must be in format 'YYYY-MM-DD HH:MM:SS'", file=sys.stderr)
                sys.exit(1)
        if args.max_runs:
            data["max_runs"] = args.max_runs
        if args.dag_id:
            data["dag_id"] = args.dag_id
        if args.image_paths:
            paths = [p.strip() for p in args.image_paths.split(",") if p.strip()]
            data["image_paths"] = paths
        if args.depends_on:
            ids = [int(x.strip()) for x in args.depends_on.split(",") if x.strip().isdigit()]
            data["depends_on"] = [
                {"task_id": tid, "inject_result": args.inject_result}
                for tid in ids
            ]

        result = api_request("POST", "/tasks", json_data=data)
        print(json.dumps(result, indent=2))

    elif args.method == "list":
        tasks = api_request("GET", "/tasks")

        if args.status:
            tasks = [t for t in tasks if t["status"] == args.status]

        if args.json:
            print(json.dumps(tasks, indent=2))
        else:
            if not tasks:
                print("No tasks found.")
            for task in tasks:
                icon = STATUS_ICONS.get(task["status"], "❓")
                dag_label = f" [dag:{task['dag_id']}]" if task.get("dag_id") else ""
                print(f"{icon} #{task['id']}: {task['title']} ({task['status']}){dag_label}")

    elif args.method == "get":
        if not args.task_id:
            parser.error("--task-id is required for get method")
        task = api_request("GET", f"/tasks/{args.task_id}")
        print(json.dumps(task, indent=2))

    elif args.method == "runs":
        if not args.task_id:
            parser.error("--task-id is required for runs method")
        runs = api_request("GET", f"/tasks/{args.task_id}/runs")
        print(json.dumps(runs, indent=2))

    elif args.method == "output":
        if not args.task_id:
            parser.error("--task-id is required for output method")
        result = api_request("GET", f"/tasks/{args.task_id}/output")

        if args.json:
            print(json.dumps(result, indent=2))
        else:
            if result["is_running"]:
                print("Status: Running")
                print("\nOutput:")
            else:
                print("Status: Not running")
            print(result["output"])

    elif args.method == "events":
        if not args.task_id:
            parser.error("--task-id is required for events method")
        params = {"limit": args.limit, "offset": args.offset}
        result = api_request("GET", f"/tasks/{args.task_id}/events", params=params)
        print(json.dumps(result, indent=2))

    elif args.method == "cancel":
        if not args.task_id:
            parser.error("--task-id is required for cancel method")
        result = api_request("POST", f"/tasks/{args.task_id}/cancel")
        print(json.dumps(result, indent=2))

    elif args.method == "retry":
        if not args.task_id:
            parser.error("--task-id is required for retry method")
        result = api_request("POST", f"/tasks/{args.task_id}/retry")
        print(json.dumps(result, indent=2))

    elif args.method == "delete":
        if not args.task_id:
            parser.error("--task-id is required for delete method")
        req = urllib.request.Request(
            f"{API_BASE}/tasks/{args.task_id}", method="DELETE"
        )
        try:
            with urllib.request.urlopen(req) as resp:
                result = json.loads(resp.read().decode("utf-8"))
                print(json.dumps(result, indent=2))
        except urllib.error.HTTPError as e:
            try:
                err_body = json.loads(e.read().decode("utf-8"))
                detail = err_body.get("error", e.reason)
            except Exception:
                detail = e.reason
            print(f"API error (HTTP {e.code}): {detail}", file=sys.stderr)
            sys.exit(1)

    elif args.method == "dag":
        if not args.dag_id:
            parser.error("--dag-id is required for dag method")
        tasks = api_request("GET", f"/dag/{args.dag_id}")
        if args.json:
            print(json.dumps(tasks, indent=2))
        else:
            print(f"DAG: {args.dag_id} ({len(tasks)} tasks)")
            for task in tasks:
                icon = STATUS_ICONS.get(task["status"], "❓")
                deps = task.get("dependencies", [])
                dep_ids = ", ".join(f"#{d['depends_on_task_id']}" for d in deps)
                dep_label = f" <- {dep_ids}" if deps else ""
                print(f"  {icon} #{task['id']}: {task['title']} ({task['status']}){dep_label}")

    elif args.method == "health":
        result = api_request("GET", "/health")
        print(json.dumps(result, indent=2))


if __name__ == "__main__":
    main()

File: skills/agentforge/scripts/agentforge_api.ts
#!/usr/bin/env bun
/**
 * AgentForge CLI - Command-line interface for AgentForge API
 *
 * Usage:
 *   agentforge_api.ts --method create --prompt "..." [options]
 *   agentforge_api.ts --method list [--status STATUS]
 *   agentforge_api.ts --method get --task-id ID
 *   agentforge_api.ts --method runs --task-id ID
 *   agentforge_api.ts --method output --task-id ID
 *   agentforge_api.ts --method events --task-id ID [--limit N]
 *   agentforge_api.ts --method cancel --task-id ID
 *   agentforge_api.ts --method retry --task-id ID
 *   agentforge_api.ts --method delete --task-id ID
 *   agentforge_api.ts --method dag --dag-id ID
 *   agentforge_api.ts --method health
 *
 * Examples:
 *   agentforge_api.ts --method create --prompt "Run tests" --title "Test run" --dir /path/to/project
 *   agentforge_api.ts --method create --prompt "Backup" --schedule delayed --delay 300
 *   agentforge_api.ts --method create --prompt "Deploy" --schedule scheduled_at --at "2026-02-15 14:30:00"
 *   agentforge_api.ts --method create --prompt "Daily report" --schedule cron --cron "0 2 * * *"
 *   agentforge_api.ts --method create --prompt "Step 2" --depends-on 10,11 --inject-result --dag-id my-pipeline
 *   agentforge_api.ts --method list --status running
 *   agentforge_api.ts --method dag --dag-id my-pipeline
 */

import { parseArgs } from "node:util";

const API_BASE = "http://127.0.0.1:9712/api";

function fail(message: string): never {
  console.error(message);
  process.exit(1);
}

async function apiRequest(
  method: string,
  path: string,
  jsonData?: unknown,
  params?: Record<string, string | number>,
): Promise<any> {
  let url = `${API_BASE}${path}`;
  if (params) {
    const qs = new URLSearchParams();
    for (const [k, v] of Object.entries(params)) qs.set(k, String(v));
    url += `?${qs.toString()}`;
  }

  let resp: Response;
  try {
    resp = await fetch(url, {
      method,
      ...(jsonData !== undefined
        ? { body: JSON.stringify(jsonData), headers: { "Content-Type": "application/json" } }
        : {}),
    });
  } catch (err: any) {
    const reason = String(err?.cause ?? err?.message ?? err);
    if (reason.includes("ECONNREFUSED") || reason.includes("Unable to connect")) {
      fail(
        "Connection failed: AgentForge is not running.\n" +
          "Start it with: cd <project-root>/backend && bun taskboard.ts",
      );
    }
    fail(`Network error: ${reason}`);
  }

  if (!resp.ok) {
    let detail: string = resp.statusText;
    try {
      const errBody: any = await resp.json();
      detail = errBody.error ?? resp.statusText;
    } catch {
      /* non-JSON error body */
    }
    fail(`API error (HTTP ${resp.status}): ${detail}`);
  }
  return resp.json();
}

const STATUS_ICONS: Record<string, string> = {
  pending: "🕐",
  scheduled: "⏰",
  running: "⏳",
  completed: "✅",
  failed: "❌",
  cancelled: "🚫",
  blocked: "⊘",
};

const METHODS = [
  "create",
  "list",
  "get",
  "runs",
  "output",
  "events",
  "cancel",
  "retry",
  "delete",
  "dag",
  "health",
] as const;

async function main(): Promise<void> {
  const { values: args } = parseArgs({
    args: process.argv.slice(2),
    options: {
      method: { type: "string" },
      prompt: { type: "string" },
      title: { type: "string" },
      dir: { type: "string", default: "." },
      schedule: { type: "string", default: "immediate" },
      cron: { type: "string" },
      delay: { type: "string" },
      at: { type: "string" },
      "max-runs": { type: "string" },
      tags: { type: "string", default: "" },
      "image-paths": { type: "string" },
      "depends-on": { type: "string" },
      "inject-result": { type: "boolean", default: false },
      "dag-id": { type: "string" },
      "task-id": { type: "string" },
      status: { type: "string" },
      limit: { type: "string", default: "1000" },
      offset: { type: "string", default: "0" },
      json: { type: "boolean", default: false },
    },
  });

  const method = args.method;
  if (!method || !(METHODS as readonly string[]).includes(method)) {
    fail(`--method is required and must be one of: ${METHODS.join(", ")}`);
  }
  if (!["immediate", "delayed", "scheduled_at", "cron"].includes(args.schedule!)) {
    fail("--schedule must be one of: immediate, delayed, scheduled_at, cron");
  }

  const taskId = args["task-id"] ? Number(args["task-id"]) : undefined;
  const requireTaskId = (m: string): number => {
    if (!taskId) fail(`--task-id is required for ${m} method`);
    return taskId;
  };

  if (method === "create") {
    if (!args.prompt) fail("--prompt is required for create method");

    const data: Record<string, unknown> = {
      prompt: args.prompt,
      title: args.title || args.prompt.slice(0, 60),
      working_dir: args.dir,
      schedule_type: args.schedule,
      tags: args.tags,
    };

    if (args.cron) data.cron_expr = args.cron;
    if (args.delay) data.delay_seconds = Number(args.delay);
    if (args.at) {
      const m = /^(\d{4})-(\d{2})-(\d{2}) (\d{2}):(\d{2}):(\d{2})$/.exec(args.at);
      if (!m) fail("Error: --at must be in format 'YYYY-MM-DD HH:MM:SS'");
      data.next_run_at = `${m[1]}-${m[2]}-${m[3]}T${m[4]}:${m[5]}:${m[6]}`;
    }
    if (args["max-runs"]) data.max_runs = Number(args["max-runs"]);
    if (args["dag-id"]) data.dag_id = args["dag-id"];
    if (args["image-paths"]) {
      data.image_paths = args["image-paths"]
        .split(",")
        .map((p) => p.trim())
        .filter(Boolean);
    }
    if (args["depends-on"]) {
      const ids = args["depends-on"]
        .split(",")
        .map((x) => x.trim())
        .filter((x) => /^\d+$/.test(x))
        .map(Number);
      data.depends_on = ids.map((tid) => ({
        task_id: tid,
        inject_result: args["inject-result"],
      }));
    }

    const result = await apiRequest("POST", "/tasks", data);
    console.log(JSON.stringify(result, null, 2));
  } else if (method === "list") {
    let tasks: any[] = await apiRequest("GET", "/tasks");

    if (args.status) tasks = tasks.filter((t) => t.status === args.status);

    if (args.json) {
      console.log(JSON.stringify(tasks, null, 2));
    } else {
      if (!tasks.length) console.log("No tasks found.");
      for (const task of tasks) {
        const icon = STATUS_ICONS[task.status] ?? "❓";
        const dagLabel = task.dag_id ? ` [dag:${task.dag_id}]` : "";
        console.log(`${icon} #${task.id}: ${task.title} (${task.status})${dagLabel}`);
      }
    }
  } else if (method === "get") {
    const id = requireTaskId("get");
    const task = await apiRequest("GET", `/tasks/${id}`);
    console.log(JSON.stringify(task, null, 2));
  } else if (method === "runs") {
    const id = requireTaskId("runs");
    const runs = await apiRequest("GET", `/tasks/${id}/runs`);
    console.log(JSON.stringify(runs, null, 2));
  } else if (method === "output") {
    const id = requireTaskId("output");
    const result = await apiRequest("GET", `/tasks/${id}/output`);

    if (args.json) {
      console.log(JSON.stringify(result, null, 2));
    } else {
      if (result.is_running) {
        console.log("Status: Running");
        console.log("\nOutput:");
      } else {
        console.log("Status: Not running");
      }
      console.log(result.output);
    }
  } else if (method === "events") {
    const id = requireTaskId("events");
    const result = await apiRequest("GET", `/tasks/${id}/events`, undefined, {
      limit: Number(args.limit),
      offset: Number(args.offset),
    });
    console.log(JSON.stringify(result, null, 2));
  } else if (method === "cancel") {
    const id = requireTaskId("cancel");
    const result = await apiRequest("POST", `/tasks/${id}/cancel`);
    console.log(JSON.stringify(result, null, 2));
  } else if (method === "retry") {
    const id = requireTaskId("retry");
    const result = await apiRequest("POST", `/tasks/${id}/retry`);
    console.log(JSON.stringify(result, null, 2));
  } else if (method === "delete") {
    const id = requireTaskId("delete");
    const result = await apiRequest("DELETE", `/tasks/${id}`);
    console.log(JSON.stringify(result, null, 2));
  } else if (method === "dag") {
    if (!args["dag-id"]) fail("--dag-id is required for dag method");
    const tasks: any[] = await apiRequest("GET", `/dag/${args["dag-id"]}`);
    if (args.json) {
      console.log(JSON.stringify(tasks, null, 2));
    } else {
      console.log(`DAG: ${args["dag-id"]} (${tasks.length} tasks)`);
      for (const task of tasks) {
        const icon = STATUS_ICONS[task.status] ?? "❓";
        const deps: any[] = task.dependencies ?? [];
        const depIds = deps.map((d) => `#${d.depends_on_task_id}`).join(", ");
        const depLabel = deps.length ? ` <- ${depIds}` : "";
        console.log(`  ${icon} #${task.id}: ${task.title} (${task.status})${depLabel}`);
      }
    }
  } else if (method === "health") {
    const result = await apiRequest("GET", "/health");
    console.log(JSON.stringify(result, null, 2));
  }
}

await main();
